streaming/base/dataset.py

```diff
@@ -659,79 +763,183 @@ def _get_work(self, world: World, epoch: int, sample_in_epoch: int) -> NDArray[n
+
+        Args:
+            shard_id (int): Shard to evict.
+        """
+        # Lock the cache. FileLocks contain threading Locks, which are not pickleable, which is
+        # incompatible with spawn, so must be created lazily.
+        if not hasattr(self, '_cache_filelock'):
+            self._cache_filelock = FileLock(self._cache_filelock_path)
+
+        # Do the eviction behind the lock.
+        with self._cache_filelock:
+            self._evict_shard(shard_id)
+
+    def evict_coldest_shard(self) -> None:
+        """Evict the coldest (i.e., least recently accessed) shard.
+
+        This method is multithread/multiprocess-safe.
+        """
+        # Lock the cache. FileLocks contain threading Locks, which are not pickleable, which is
+        # incompatible with spawn, so must be created lazily.
+        if not hasattr(self, '_cache_filelock'):
+            self._cache_filelock = FileLock(self._cache_filelock_path)
+
+        # Do the eviction behind the lock.
+        with self._cache_filelock:
+            self._evict_coldest_shard()
+
+    def download_shard(self, shard_id: int, blocking: bool = True) -> None:
+        """Download a shard, either waiting or skipping if in progress by another worker.
+
+        This method is multithread/multiprocess-safe.
+
+        If cache limit is enabled, this method may delete one or more other shards to make space
+        for this download.
+
+        Args:
+            shard_id (int): Shard to download.
+            blocking (bool): Whether to wait or skip if the shard is currently being downloaded by
+                someone else.
+        """
+        # Lock the cache. FileLocks contain threading Locks, which are not pickleable, which is
+        # incompatible with spawn, so must be created lazily.
+        if not hasattr(self, '_cache_filelock'):
+            self._cache_filelock = FileLock(self._cache_filelock_path)
+        lock = self._cache_filelock
         lock.acquire()
-        state = shard_states[shard_id]
-        if state == _ShardState.UNKNOWN:
-            shard_states[shard_id] = _ShardState.DOWNLOADING
-            lock.release()
+
+        # Get the state of the shard to download.
+        state = self._shard_states[shard_id]
+
+        if state == _ShardState.MISSING:
+            # If missing, transition state to downloading.
+            self._shard_states[shard_id] = _ShardState.DOWNLOADING
+
+            # Get the stream and shard.
             stream_id = self.stream_per_shard[shard_id]
             stream = self.streams[stream_id]
             shard = self.shards[shard_id]
+
+            # If cache_limit is enabled, we first may have to make space for the new shard.
+            if self.cache_limit:
+                # Here, we give the shard a provisional access time so it doesn't get inadvertently
+                # evicted if we need to evict shard(s).
+                self._shard_access_times[shard_id] = time()
```

Comment:
at this point, this shard is missing, so its download time should be never, right? This line is safe, just checking my understanding because I think it is unnecessary.